src/order/heyting_algebra.lean

```diff
@@ -0,0 +1,678 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import order.bounded_order
+
+/-!
+# Heyting algebras
+
+This file defines Heyting algebras.
+
+## Main declarations
+
+* `generalized_heyting_algebra`
+* `generalized_coheyting_algebra`
+* `heyting_algebra`
+* `coheyting_algebra`
+* `biheyting_algebra`
+
+## References
+
+* [Francis Borceux, *Handbook of Categorical Algebra III*][borceux-vol3]
+-/
+
+set_option old_structure_cmd true
+
+open function order_dual
+
+universes u
+variables {ι α β : Type*}
+
+/-! ### Notation -/
+
+/-- Syntax typeclass for Heyting implication. -/
+@[notation_class] class has_himp (α : Type*) := (himp : α → α → α)
+
+/-- Syntax typeclass for Heyting negation. -/
+@[notation_class] class has_hnot (α : Type*) := (hnot : α → α)
+
+export has_himp (himp) has_sdiff (sdiff) has_hnot (hnot)
+
+infixr ` ⇨ `:60 := himp
+prefix `￢`:72 := hnot
+
+instance [has_himp α] [has_himp β] : has_himp (α × β) := ⟨λ a b, (a.1 ⇨ b.1, a.2 ⇨ b.2)⟩
+instance [has_hnot α] [has_hnot β] : has_hnot (α × β) := ⟨λ a, (￢a.1, ￢a.2)⟩
+instance [has_sdiff α] [has_sdiff β] : has_sdiff (α × β) := ⟨λ a b, (a.1 \ b.1, a.2 \ b.2)⟩
+instance [has_compl α] [has_compl β] : has_compl (α × β) := ⟨λ a, (a.1ᶜ, a.2ᶜ)⟩
+
+@[simp] lemma fst_himp [has_himp α] [has_himp β] (a b : α × β) : (a ⇨ b).1 = a.1 ⇨ b.1 := rfl
+@[simp] lemma snd_himp [has_himp α] [has_himp β] (a b : α × β) : (a ⇨ b).2 = a.2 ⇨ b.2 := rfl
+@[simp] lemma fst_hnot [has_hnot α] [has_hnot β] (a : α × β) : (￢a).1 = ￢a.1 := rfl
+@[simp] lemma snd_hnot [has_hnot α] [has_hnot β] (a : α × β) : (￢a).2 = ￢a.2 := rfl
+@[simp] lemma fst_sdiff [has_sdiff α] [has_sdiff β] (a b : α × β) : (a \ b).1 = a.1 \ b.1 := rfl
+@[simp] lemma snd_sdiff [has_sdiff α] [has_sdiff β] (a b : α × β) : (a \ b).2 = a.2 \ b.2 := rfl
+@[simp] lemma fst_compl [has_compl α] [has_compl β] (a : α × β) : aᶜ.1 = a.1ᶜ := rfl
+@[simp] lemma snd_compl [has_compl α] [has_compl β] (a : α × β) : aᶜ.2 = a.2ᶜ := rfl
+
+namespace pi
+variables {π : ι → Type*}
+
+instance [Π i, has_himp (π i)] : has_himp (Π i, π i) := ⟨λ a b i, a i ⇨ b i⟩
+instance [Π i, has_hnot (π i)] : has_hnot (Π i, π i) := ⟨λ a i, ￢a i⟩
+
+lemma himp_def [Π i, has_himp (π i)] (a b : Π i, π i) : (a ⇨ b) = λ i, a i ⇨ b i := rfl
+lemma hnot_def [Π i, has_hnot (π i)] (a : Π i, π i) : ￢a = λ i, ￢a i := rfl
+
+@[simp] lemma himp_apply [Π i, has_himp (π i)] (a b : Π i, π i) (i : ι) : (a ⇨ b) i = a i ⇨ b i :=
+rfl
+@[simp] lemma hnot_apply [Π i, has_hnot (π i)] (a : Π i, π i) (i : ι) : (￢a) i = ￢a i := rfl
+
+end pi
+
+/-- A generalized Heyting algebra is a lattice with an additional binary operation `⇨` called
+Heyting implication such that `a ⇨` is right adjoint to `a ⊓`. -/
+class generalized_heyting_algebra (α : Type*) extends lattice α, has_top α, has_himp α :=
+(le_top : ∀ a : α, a ≤ ⊤)
```

Comment:
Why not extend `order_top`?

Reply:
It's a mixin, so that doesn't work. See `complete_lattice` or `boolean_algebra`:
https://github.com/leanprover-community/mathlib/blob/24a55a0540a29f4b1afcb0cc8f716c86b2a47a9f/src/order/boolean_algebra.lean#L589-L592